packages/explat-client/src/create-explat-client.ts

```diff
@@ -0,0 +1,199 @@
+}
+
+export class MissingExperimentAssignmentError extends Error {
+	constructor( ...params: any[] ) {
+		super( ...params );
+
+		// Maintains proper stack trace for where our error was thrown (only available on V8)
+		if ( Error.captureStackTrace ) {
+			Error.captureStackTrace( this, MissingExperimentAssignmentError );
+		}
+
+		this.name = 'MissingExperimentAssignmentError';
+	}
+}
+
+/**
+ * Create an ExPlat Client
+ *
+ * @param config Configuration object
+ */
+export default function createExPlatClient( config: Config ): ExPlatClient {
+	if ( typeof window === 'undefined' ) {
+		throw new Error( 'Running outside of a browser context.' );
+	}
+
+	const store = State.createStore();
+
+	/**
+	 * This bit of code is the heavy lifting behind loadExperimentAssignment, allowing it to be used intuitively.
+	 *
+	 * Using asyncOneAtATime, is how we ensure for each experiment that there is only ever one fetch process occuring.
+	 *
+	 *
+	 * @param experimentName The experiment's name
+	 */
+	const createWrappedExperimentAssignmentFetchAndStore = ( experimentName: string ) =>
+		Timing.asyncOneAtATime( async () => {
+			const fetchedExperimentAssignment = await Request.fetchExperimentAssignment(
+				config,
+				experimentName
+			);
+			State.storeExperimentAssignment( store, fetchedExperimentAssignment );
+			return fetchedExperimentAssignment;
+		} );
+	const experimentNameToWrappedExperimentAssignmentFetchAndStore: Record<
+		string,
+		() => Promise< ExperimentAssignment >
+	> = {};
+
+	return {
+		loadExperimentAssignment: async ( experimentName: string ): Promise< ExperimentAssignment > => {
+			try {
+				if ( ! Validation.isName( experimentName ) ) {
+					throw new Error( `Invalid experimentName: ${ experimentName }` );
+				}
+
+				const storedExperimentAssignment = State.retrieveExperimentAssignment(
+					store,
+					experimentName
+				);
+				if (
+					storedExperimentAssignment &&
+					ExperimentAssignments.isAlive( storedExperimentAssignment )
+				) {
+					return storedExperimentAssignment;
+				}
+
+				if (
+					experimentNameToWrappedExperimentAssignmentFetchAndStore[ experimentName ] === undefined
+				) {
+					experimentNameToWrappedExperimentAssignmentFetchAndStore[
+						experimentName
+					] = createWrappedExperimentAssignmentFetchAndStore( experimentName );
+				}
+				// We time out the request here and not above so the fetch-and-store continues and can be
+				// returned by future uses of loadExperimentAssignment.
+				const fetchedExperimentAssignment = await Timing.timeoutPromise(
+					experimentNameToWrappedExperimentAssignmentFetchAndStore[ experimentName ](),
+					EXPERIMENT_FETCH_TIMEOUT
+				);
+				if ( ! fetchedExperimentAssignment ) {
+					throw new Error( `Could not fetch ExperimentAssignment` );
+				}
+
+				return fetchedExperimentAssignment;
+			} catch ( e ) {
+				// We need to be careful about errors thrown in the catch block too:
```

Comment:
My idea here is to make this failsafe, and the second try-catch makes that simpler: if we have any code here we should assume it can go wrong and have a backup plan. If we push any failures into the backup plan we then only need to ensure the backup plan itself can't go wrong rather than the rest of the code.

In this case the backup plan is the final `createFallbackExperimentAssignment`.


> Lines 134-148 shouldn't throw exceptions, should they?

They shouldn't but now we are covered in the case they do.

Also it is not just `logError` but also `storeExperimentAssignment` - the internals can also throw.

Reply:
OK, it seems a bit too defensive to me, but I still think that flattening the try-catches is worthwhile. Having three nested levels of try-catch blocks with error names like `e`, `e2`, and `e3` is harder to follow than the flow I suggested, which makes it clear that we only get into fallbacks outside dev mode.